Various minor fixes.

brett19 opened the pull request:
The commit messages should explain everything.

brett19 committed:
Fixed renderdoc crash due to bindings which are not tightly packed

In the case where bindings are not tightly packed 1:1 with the
attributes, the bindings weren't being resolved first.

brett19 committed:
Added missing tracking for patching.

This caused a crash in development builds due to the
assertion immediately afterwards failing.

baldurk commented on renderdoc/driver/vulkan/vk_overlay.cpp:
```diff
@@ -308,6 +308,8 @@ void VulkanDebugManager::PatchFixedColShader(VkShaderModule &mod, float col[4])
       else
         RDCERR("Unexpected constant value");
+
+      patched++;
```
Good catch here, I completely messed up on this check (not sure how I missed that). The increment of `patched` though should happen inside the if statements where the constants are actually patched. The intention was to catch the case where somehow less than 4 constants were found so the patching didn't succeed but there wasn't an "Unexpected constant value" error.

brett19 replied:
I figured since the else block was an error, and all the if blocks actually did patching, it made more sense to put it in one place and check that at least the outer if inside the loop caught 4 cases.
P.S. I had no clue what this is actually doing, I just took a random stab at how it was meant to look.

baldurk replied:
Yeh actually that's fair enough, either way there is an error indicating that something went wrong.

baldurk commented on renderdoc/driver/vulkan/vk_postvs.cpp:
```diff
@@ -56,6 +56,21 @@ static void ConvertToMeshOutputCompute(const ShaderReflection &refl, const SPIRV
   SPIRVEditor editor(modSpirv);
 
+  // strip away any execution modes from the original
+  for (SPIRVIterator it = editor.BeginAll(); it; it++) {
```
I don't think this new function is needed. The spec has a clear layout for where things are expected to happen, and `OpExecutionMode` opcodes only happen after entry points and before debug instructions.

I don't have each section precisely but you should be able to iterate from `BeginEntries` to `BeginDebug` and that will cover the area where execution modes could be declared.

brett19 replied:
Oh!  I spent a while trying to figure out how I was supposed to access those sections, since there were not obvious helpers to fetch those areas, and there was no obvious way to loop everything.  Will fix that in a moment.

baldurk commented on renderdoc/driver/vulkan/vk_postvs.cpp:
```diff
@@ -56,6 +56,21 @@ static void ConvertToMeshOutputCompute(const ShaderReflection &refl, const SPIRV
 
+  // strip away any execution modes from the original
+  for (SPIRVIterator it = editor.BeginAll(); it; it++) {
+     if (it.opcode() == spv::OpExecutionMode) {
```
Removing all execution modes unconditionally is not always the right thing to do. We don't know what the execution mode is and it may be necessary to run correctly - which means either we need to translate it/process it for compute, or it may be something that's valid for all shader types.

From your message on IRC it sounded like you maybe had multiple entry points in your SPIR-V module and since this patching process removed all other entry points you ended up with a dangling `OpExecutionMode` pointing to an invalid ID. That's a narrower definition of things we need to strip - `OpExecutionMode` entries that point to other entry points.

brett19 committed:
Fix ExecutionMode's crashing drivers with compute based mesh output.

The OpExecutionMode entries were being left in place, but pointing
to functions which are no longer entry points.  This caused the GPU
driver to crash (probably trying to look it up).

baldurk commented on renderdoc/driver/vulkan/vk_postvs.cpp:
```diff
@@ -659,6 +659,38 @@ static void ConvertToMeshOutputCompute(const ShaderReflection &refl, const SPIRV
+  // Strip away any execution modes from the original shaders
+  for (SPIRVIterator it = editor.BeginEntries(), end = editor.BeginDebug(); it < end; ++it) {
+     if (it.opcode() == spv::OpExecutionMode) {
+        auto modeEntryID = static_cast<SPIRVId>(it.word(1));
```
Avoid `static_cast` and use of `auto` for simple types please,

```suggestion
        SPIRVId modeEntryID = SPIRVId(it.word(1));
```

Same below for `execMode`.
